KAFKA-5987: Maintain order of metric tags in generated documentation

rhauch opened the pull request:
The `MetricNameTemplate` is changed to used a `LinkedHashSet` to maintain the same order of the tags that are passed in. This tag order is then maintained when `Metrics.toHtmlTable` generates the MBean names for each of the metrics.

The `SenderMetricsRegistry` and `FetcherMetricsRegistry` both contain templates used in the producer and consumer, respectively, and these were changed to use a `LinkedHashSet` to maintain the order of the tags.

Before this change, the generated HTML documentation might use MBean names like the following and order them:

```
kafka.connect:type=sink-task-metrics,connector={connector},partition={partition},task={task},topic={topic}
kafka.connect:type=sink-task-metrics,connector={connector},task={task}
```
However, after this change, the documentation would use the following order:
```
kafka.connect:type=sink-task-metrics,connector={connector},task={task}
kafka.connect:type=sink-task-metrics,connector={connector},task={task},topic={topic},partition={partition}
```

This is more readable as the code that is creating the templates has control over the order of the tags.

Note that JMX MBean names use ObjectName that does not maintain order of the properties (tags), so this change should have no impact on the actual JMX MBean names used in the metrics.

cc @wushujames

rhauch committed:
KAFKA-5987 Maintain order of metric tags in generated documentation

The `MetricNameTemplate` is changed to used a `LinkedHashSet` to maintain the same order of the tags that are passed in. This tag order is then maintained when `Metrics.toHtmlTable` generates the MBean names for each of the metrics.

The `SenderMetricsRegistry` and `FetcherMetricsRegistry` both contain templates used in the producer and consumer, respectively, and these were changed to use a `LinkedHashSet` to maintain the order of the tags.

wushujames commented on clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java:
```diff
@@ -26,27 +26,28 @@
-    private Set<String> tags;
+    private LinkedHashSet<String> tags;
 
     public MetricNameTemplate(String name, String group, String description, Set<String> tags) {
```
Someone could construct a HashSet that they added items to in the order "a" "b" "c", and pass it into MetricNameTemplate. It might then get manged here into the order "c" "b" "a". Should we have the MetricNameTemplate accept a LinkedHashSet(), instead of a Set?

wushujames replied:
Doing this would also make the compile complain about all instances that instantiate MetricNameTemplates with a non-order-preserving Set, which I think is what we want. As it currently is implemented in the PR, I don't think the compile actually catches anything.

rhauch replied:
Yes, it is possible at the moment to not provide an order-preserving set. However, that change snowballs to places where, for example, [KafkaConsumer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L663) creates a singleton map for one tag and value, and then passes the `keySet()` from that map. Even when the map is a `LinkedHashMap`, the `keySet()` method just returns a normal `Set`. Requiring a `LinkedHashSet` makes everything harder to use.

I agree that it is not as desirable to pass in just a `Set` into the `MetricNameTemplate`, but this way it is up to the callers to decide whether they want their tags ordered. I'm open to changing this, but this will involve more changes.

wushujames replied:
I agree that enforcing a LinkedHashSet is harder to use. I'm a little worried about discoverability, though. As it is, there is no way to "discover" the "feature" that passing in an order-preserving map will, in fact, preserve order. It kind of feels obvious that it *should*, but it also kind of doesn't, since the constructor signature doesn't say anything about that.

Maybe add a comment to the javadocs for MetricNameTemplate() that if you pass in an order-preserving-Map, that it will preserve order? Or is that a useless comment, because it'll be easy to miss? As it is, the "ability to preserve order" is kind of an optional "power-user" feature, so maybe hiding the functionality in a comment is okay.

In your callers, maybe add a comment of "passing in a LinkedHashSet, so that order of tags is preserved in the docs"?

rhauch replied:
Yeah, I think documenting it is better than making it much harder to use. Really, I think the only thing it affects is the generated documentation, and if somebody sees the order as wrong they (hopefully) will check the `MetricNameTemplate` JavaDoc.

rhauch committed:
KAFKA-5987: Added more documentation for the MetricNameTemplate and HTML table generation

rhauch commented:
@wushujames, I've updated the JavaDoc to hopefully make things more explicit. WDYT?

wushujames commented on clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java:
```diff
@@ -26,47 +26,86 @@
-        this(name, group, description, getTags(keys));
+
+    /**
+     * Create a new template. Note that the order of the tags will be preserved if the supplied
```
The sentence "Note that the order of the tags..." is not applicable because there is no set in this case. It's just the order of the varargs.

rhauch replied:
Fixed.

rhauch committed:
KAFKA-5987: Minor correction to JavaDoc

wushujames commented on clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java:
```diff
@@ -26,47 +26,86 @@
+     * @param name the name of the metric; may not be null
+     * @param group the name of the group; may not be null
+     * @param description the description of the metric; may not be null
+     * @param tagsNames the set of metric tag names, which can/should be a set that maintains order; may not be null
```
Interesting. If someone passes in a non-order-preserving Set, then does that mean that the order of the MBeans in the docs might change from one build to the next? The build of the docs won't be repeatable, then?

rhauch replied:
Technically, it is possible that the order is different on different JVMs.

rhauch replied:
BTW, this is true even without the PR.

wushujames replied:
Kinda, but it had a TreeMap before, so it used the "natural ordering" of the Strings which means... something.

Anyway, are we okay with the fact that the ordering may change according to the person who built the docs? The kafka.apache.org webpage will then show whatever order the site-builder's JVM used? Or is it Jenkins?

We don't really have any other choice, other than making everyone use LinkedHashSets, right? And we decided we don't want that.

rhauch replied:
True that before it used `TreeMap`. I guess I'd like to get some other feedback, perhaps from @ijuma and @rajinisivaram about whether it'd be better to make all the changes to enforce order.

ewencp replied:
Unfortunately this is public API, so just changing it across the board may not be trivial and requires a KIP.

I'd suggest getting the quick documentation win via this PR (modulo my other comment) and deal with any more extensive change in a follow up JIRA.

rhauch commented:
@ijuma and @rajinisivaram, what do you think about the proposed changes to the metrics documentation generator and the discussion above?

rajinisivaram commented:
@rhauch Personally, I would prefer to make changes everywhere to enforce order. Even though it means a bigger change, it becomes much more obvious, especially if a new group of tags is added in future. Perhaps wait and see what @ijuma thinks.

ijuma commented:
Sorry for the delay. @rhauch, I'm not sure I understand:

"Note that JMX MBean names use ObjectName that does not maintain order of the properties (tags), so this change should have no impact on the actual JMX MBean names used in the metrics."

Shouldn't we make sure that the JMX mbean names match the documentation?

ewencp commented:
@ijuma I'm not sure there are enough guarantees to make that possible, of `ObjectName` loses order there's not much we can do to match, is there?

ewencp commented on clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java:
```diff
@@ -26,47 +26,85 @@
+     *
+     * @return the ordered set of tag names; never null but possibly empty
+     */
+    public LinkedHashSet<String> tags() {
```
Perhaps we should not change the return type here unless we decide to make the more extensive `LinkedHashSet` change across all APIs (and corresponding KIP).

rhauch replied:
Agreed, this can be reverted back to return `Set<String>`. FWIW, this PR started out in the same release as the code that added this class, so changing it then wouldn't have been a problem; changing it now would require a KIP.

rhauch committed:
KAFKA-5987: Reverted return type of MetricNameTemplate.tags()

Unable to change this public API without a KIP, so reverting. Only uses are as a basic collection anyway, and this change was to simply reinforce the tighter sematics.

ewencp reviewed (approved):
LGTM, merging to trunk, 1.1, and 1.0
